src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -0,0 +1,431 @@
+import Data.List   (sort)
+import Data.Tuple  (swap)
+import GHC.Generics
+
+import qualified Algebra.Graph              as G
+import qualified Algebra.Graph.AdjacencyMap as AM
+
+import qualified Data.Map.Strict as Map
+import qualified Data.Set        as Set
+
+{-| The 'Bipartite.AdjacencyMap' data type represents an __undirected__
+bipartite graph by two maps of vertices into their adjacency sets. The two type
+parameteters define the types of identifiers of the vertices of each part.
+
+__Note:__ even if the identifiers and their types for two vertices of
+different parts are equal, these vertices are considered to be different.
+See examples for more details.
+-}
+data AdjacencyMap a b = BAM {
+    -- | The /adjacency map/ of the left part of the graph: each vertex is
+    -- associated with a set of its neighbours. Complexity: /O(1)/ time and
+    -- memory.
+    --
+    -- @
+    -- leftAdjacencyMap 'empty'                    == Map.'Map.empty'
+    -- leftAdjacencyMap ('leftVertex' 1)           == Map.'Map.singleton' 1 Set.'Set.empty'
+    -- leftAdjacencyMap ('rightVertex' 1)          == Map.'Map.empty'
+    -- leftAdjacencyMap ('edge' 1 2)               == Map.'Map.singleton' 1 (Set.'Set.singleton' 2)
+    -- leftAdjacencyMap ('edge' 1 1)               == Map.'Map.singleton' 1 (Set.'Set.singleton' 1)
+    -- leftAdjacencyMap ('edges' [(1, 1), (1, 2)]) == Map.'Map.singleton' 1 (Set.'Set.fromAscList' [1, 2])
+    -- @
+    leftAdjacencyMap :: Map.Map a (Set.Set b),
+
+    -- | The inverse map for 'leftAdjacencyMap'. Complexity: /O(1)/ time and memory.
+    --
+    -- @
+    -- rightAdjacencyMap 'empty'                    == Map.'Map.empty'
+    -- rightAdjacencyMap ('leftVertex' 1)           == Map.'Map.empty'
+    -- rightAdjacencyMap ('rightVertex' 1)          == Map.'Map.singleton' 2 Set.'Set.empty'
+    -- rightAdjacencyMap ('edge' 1 2)               == Map.'Map.singleton' 2 (Set.'Set.singleton' 1)
+    -- rightAdjacencyMap ('edge' 1 1)               == Map.'Map.singleton' 1 (Set.'Set.singleton' 1)
+    -- rightAdjacencyMap ('edges' [(1, 1), (1, 2)]) == Map.'Map.fromAscList' [(1, Set.'Set.singleton' 1), (2, Set.'Set.singleton' 1)]
+    -- @
+    rightAdjacencyMap :: Map.Map b (Set.Set a)
+} deriving (Eq, Show, Generic)
+
+-- | Check that the internal graph representation is consistent, i.e. that all
+-- edges that are present in the 'leftAdjacencyMap' are present in the
+-- 'rightAdjacencyMap' map.
+--
+-- @
+-- consistent 'empty'            == True
+-- consistent ('vertex' x)       == True
+-- consistent ('edge' x y)       == True
+-- consistent ('edges' xs)       == True
+-- consistent ('fromGraph' g)    == True
+-- consistent ('toBipartite' am) == True
+-- @
+consistent :: (Ord a, Ord b) => AdjacencyMap a b -> Bool
+consistent (BAM lr rl) = internalEdgeList lr == sort (map swap $ internalEdgeList rl)
+
+-- The list of edges of a bipartite adjacency map.
+--
+-- /Note: this function is for internal use only./
```

Comment:
This comment is redundant: all non-exported functions are internal (and one can immediately see that this function is not exported, because it doesn't have a Haddock comment, by convention).

Reply:
Whoops, forgot about it.